### DISCUSSION TITLE
Motor beep Problems reading from arduino serial inputs

### DISCUSSION THREAD

**Discussion by hanqiuli** [bitcraze/discussions#1042] [Q&A]
Hi all, I'm trying to write a simple application that reads an input character from an arduino, and if that character is "o", turn the motors on for a couple of seconds.

This is my implementation (I just copied the uart2test.c and made some minor modifications)
```

#define DEBUG_MODULE "ARDUINO"

#include <stdint.h>
#include <string.h>

#include "FreeRTOS.h"
#include "task.h"

#include "stm32fxxx.h"
#include "system.h"
#include "config.h"
#include "debug.h"
#include "deck.h"
#include "uart2.h"
#include "motors.h"

//Hardware configuration
static bool isInit;

void uart2testTask(void* arg);

static void uart2testInit(DeckInfo *info)
{
  if(isInit)
    return;

  uart2Init(19200);

  xTaskCreate(uart2testTask, UART2_TEST_TASK_NAME, UART2_TEST_TASK_STACKSIZE, NULL, UART2_TEST_TASK_PRI, NULL);

  DEBUG_PRINT("Hello Crazyflie 2.1 deck world!\n");
  isInit = true;
}

static bool uart2testTest()
{
  bool status = true;

  if(!isInit)
    return false;

  return status;
}

void uart2testTask(void* arg)
{
  systemWaitStart();

  while (1)
  { 
    char c;
    uart2Getchar(&c);
    DEBUG_PRINT("%c", c);
    if (c == 'o'){
        DEBUG_PRINT("I GOT HERE!");
        motorsTest();
    }
  }
}

static const DeckDriver arduino_deck = {
//  .vid = 0xBC,
//  .pid = 0x08,
  .name = "arduino_UART2",

  .usedGpio = DECK_USING_UART2 ,
  .init = uart2testInit,
  .test = uart2testTest,
};

DECK_DRIVER(arduino_deck);

```
I've managed to get some signals going into the crazyflie. There seems to be some issues with the UART communication, since sometimes the debug prints are not shown in console until I have sent 3 characters (then outputs all three debug prints at once). Also, the code seems to be getting into the if statement, but I have no response from the motors.

I've tried increasing the task priority and stack size, but also no luck

**Comment by hanqiuli** [comment]
```
#define DEBUG_MODULE "ARDUINO"

#include <stdint.h>
#include <string.h>

#include "FreeRTOS.h"
#include "task.h"

#include "stm32fxxx.h"
#include "system.h"
#include "config.h"
#include "debug.h"
#include "deck.h"
#include "uart2.h"
#include "motors.h"
#include "param_logic.h"


//Hardware configuration
static bool isInit;

void uart2testTask(void* arg);

static void uart2testInit(DeckInfo *info)
{
  if(isInit)
    return;

  uart2Init(19200);

  xTaskCreate(uart2testTask, UART2_TEST_TASK_NAME, UART2_TEST_TASK_STACKSIZE, NULL, UART2_TEST_TASK_PRI, NULL);

  isInit = true;
}

static bool uart2testTest()
{
  bool status = true;

  if(!isInit)
    return false;

  return status;
}

void uart2testTask(void* arg)
{
  systemWaitStart();
  uint8_t c;
  bool char_received;
   
  paramVarId_t enabled = paramGetVarId((const char*)"motorPowerSet", (const char*)"enable");
  paramVarId_t motorPower1 = paramGetVarId("motorPowerSet", "m1");
  // paramVarId_t motorPower2 = paramGetVarId("motorPowerSet", "m2");
  // paramVarId_t motorPower3 = paramGetVarId("motorPowerSet", "m3");
  // paramVarId_t motorPower4 = paramGetVarId("motorPowerSet", "m4");
  while (1)
  { 
    char_received = uart2GetCharWithDefaultTimeout(&c); // Read a character with default timeout

    if (char_received) {
        if (c == 'o') {
            vTaskDelay(M2T(10));
            paramSetInt(enabled, 1);
            vTaskDelay(M2T(10));
            paramSetInt(motorPower1, 20000);
            vTaskDelay(M2T(2000));
            paramSetInt(motorPower1, 0);
            vTaskDelay(M2T(10));
        } 
        else if (c == 'm'){
            vTaskDelay(M2T(1000));
            motorsPlayMelody(happyBirthdayMelodyHigherAndFaster);
            vTaskDelay(M2T(1000));
            DEBUG_PRINT("Melody played\n");
        }
        else {
            DEBUG_PRINT("Received char: %c\n", c); // Print the received character
        }
    }

    vTaskDelay(M2T(10)); // Delay for a short period (e.g., 10 ms)
  }
}

static const DeckDriver arduino_deck = {
//  .vid = 0xBC,
//  .pid = 0x08,
  .name = "arduino_UART2",
  .usedPeriph = DECK_USING_UART2,
  .init = uart2testInit,
  .test = uart2testTest,
};

DECK_DRIVER(arduino_deck);

```

I've managed to solve the previous issue! However I'm struggling to control the motors with the UART signals (I assume it is being interrupted by some other module). I've found a workaround by just setting the parameters (just like from the client) instead of using the motorSetRatio function from motors.c

However, I'm trying to make it play a melody now. I've managed to make the melody play at startup (by changing the motorTest function in motors.c) but it doesn't work on command. There is some clicking from the motor going on, and printing the debug messages I can see that the motorPlayMelody function is being called. 

Can anyone help me identify which part of the firmware I have to modify to override whatever's surpressing the motors? So far I've tried:
- Increasing task priority
- Editing the supervisorAreMotorsAllowedToRun function in the supervisor.c file (by just swapping it out with a return true)

**Reply by knmcguire** [reply]
hi! good that you managed to figure out the debugging. I would have indeed advised you to add the '/n' character but you already figured that one out :).

Alright, so the current problem is, that the deck driver receives the uart output correctly looking at debug messages, but the crazyflie's motors do not respond with a melody? That is interesting and perhaps indeed the supervisor has something to do with it (but I'm not exactly sure). I'll try to recreate your problem with an app-layer driver and parameter instead, as it seems to be not an uart issue anymore. I'll update once I can confirm your issue.

**Comment by knmcguire** [comment]
So yes, I've done an app layer variant where the startup motor tune should be running every 5 seconds and it does not work... perhaps something is locked after the initialization that it will only respond to rpm... but anyway, know you know that your observation is valid. 


I'll need to refer this to my colleagues as I don't really have an idea of where to look for, but at least here the app layer I've written

```
#include <string.h>
#include <stdint.h>
#include <stdbool.h>

#include "app.h"

#include "FreeRTOS.h"
#include "task.h"
#include "motors.h"
#include "sound.h"

#define DEBUG_MODULE "APP"
#include "debug.h"

#define MOTORS_TIM_BEEP_CLK_FREQ  (84000000L / 5)
const uint16_t sound[NBR_OF_MOTORS] = {A4, A5, F5, D5 };
#define MOTOR_M1  0
#define MOTOR_M2  1
#define MOTOR_M3  2
#define MOTOR_M4  3

const uint32_t MOTORSAPP[] = { MOTOR_M1, MOTOR_M2, MOTOR_M3, MOTOR_M4 };

void appMain() {
  DEBUG_PRINT("Hello! I have a motor melody app\n");

  while(1) {
    vTaskDelay(M2T(5000));

    DEBUG_PRINT("Motor melody!\n");
    int i;

    for (i = 0; i < sizeof(MOTORSAPP) / sizeof(*MOTORSAPP); i++)
    {    
      motorsBeep(MOTORSAPP[i], true, sound[i], (uint16_t)(MOTORS_TIM_BEEP_CLK_FREQ / A4)/ 20);
      vTaskDelay(M2T(MOTORS_TEST_ON_TIME_MS));
      motorsBeep(MOTORSAPP[i], false, 0, 0);
      vTaskDelay(M2T(MOTORS_TEST_DELAY_TIME_MS));
    }

  }
}
```

**Reply by tobbeanton** [reply]
I think what is happening is that the stabilizer task overwrites with [new motor values](https://github.com/bitcraze/crazyflie-firmware/blob/dbb9df1137f11d4e7e3771c56d25a7137b5b69cc/src/modules/src/stabilizer.c#L323C2-L323C2). So this has to be stopped somehow, by e.g. commenting it out but it will never fly then.
